src/runtime/recipe/particle.ts

```diff
@@ -183,9 +205,14 @@ export class Particle {
     return connection;
   }
 
-  addConnectionName(name) {
+  addConnectionName(name: string): HandleConnection {
     assert(this._connections[name] == undefined);
     this._connections[name] = new HandleConnection(name, this);
+    if (this.spec) {
+      const connSpec = this.spec.getConnectionByName(name);
+      this._connections[name].type = connSpec.type;
+      this._connections[name].direction = connSpec.direction;
```

Comment:
we should only ever update recipe information based on recipe directives, and check that against the spec as part of validation / resolution.

Reply:
I think, I need to talk to you more about this to make sure I understand.

For example (1):
```
particle P
  out Thing thing

recipe
  create as handle
  P
    thing = handle
```
do you mean that the particle-spec’s connection’s spec direction is “out” 
but the recipe’s handle-connection direction is “inout” 
and `HandleConnection.isValid` verifies that they are compatible?
does the isValid check happen only during normalization, or also during parsing? 

what if it is (2): 
```
particle P
  out Thing thing

recipe
  use as handle
  P
    thing -> handle
```
so particle-spec connection’s spec direction is “out”
but recipe handle connection direction is “in”. When does the error occur?

And in this case (3):
```
particle P &doP
  out Thing thing

recipe
  use as handle
  &doP
    thing -> handle
```
this will simply mean that particle P cannot be used here for verb &doP, right?

But then (4):
```
particle P &doP
  out Thing thing

recipe
  &doP
```
is particle P considered suitable for &doP in the recipe? And later when the handle’s connection is being created, what direction is it assigned? 
Or shall we leave `handleConnection._direction` undefined and have the `get direction` fallback to the spec, if available?

You mentioned that slot connections are verified against the spec at validation time, could you, please, point me to that code?

Reply:
You have your handle directions backwards (`thing -> handle` says thing writes _out_ to handle) but assuming they're reversed:

(1) correct, isValid verifies they're compatible. This check happens during normalization only.
(2) this is not a parser error (there's no syntactic problems with this recipe), nor is it a semantic error. It's simply an invalid recipe. So isValid will report false.
(3) Correct, when evaluating doP for matching, the strategy that substitutes verbs should note that they're incompatible and refuse to perform the substitution. If it doesn't, this is simply a validity error, which *phew* is lucky, because otherwise we'd have created a back-door into getting around parse-time errors.
(4) Yes, of course. doP is completely suitable here. When the handle connection is connected to something, then whatever is doing the connection will need to specify the connection direction (it'll default to inout). This matches with recipes defining connections: they specify a direction as part of that connection.

Slot connections don't yet appear to be verified against the spec at validation time. There's a TODO in place, but it isn't very clear. I'll talk to Alice about at least improving that TODO.

Reply:
done (i think). 
could you, please, take another look, when you get a minute?